gammapy/maps/wcsnd.py

```diff
@@ -495,38 +498,116 @@ def plot(self, ax=None, fig=None, add_cbar=False, stretch='linear', **kwargs):
 
-        caxes = ax.imshow(data, **kwargs)
 
-        cbar = fig.colorbar(caxes, ax=ax) if add_cbar else None
-        try:
-            ax.coords['glon'].set_axislabel('Galactic Longitude')
-            ax.coords['glat'].set_axislabel('Galactic Latitude')
-        except KeyError:
-            ax.coords['ra'].set_axislabel('Right Ascension')
-            ax.coords['dec'].set_axislabel('Declination')
-        except AttributeError:
-            log.info("Can't set coordinate axes. No WCS information available.")
-
-        # without this the axis limits are changed when calling scatter
-        ax.autoscale(enable=False)
-        return fig, ax, cbar
+
+        if self.geom.ndim == 2:
+            if ax is None:
+                ax = fig.add_subplot(1, 1, 1, projection=self.geom.wcs)
+
+            data = self.data
+            norm = simple_norm(data[np.isfinite(data)], stretch)
+            kwargs.setdefault('norm', norm)
+            caxes = ax.imshow(data, **kwargs)
+            cbar = fig.colorbar(caxes, ax=ax) if add_cbar else None
+            # without this the axis limits are changed when calling scatter
+            ax.autoscale(enable=False)
+
+            try:
+                ax.coords['glon'].set_axislabel('Galactic Longitude')
+                ax.coords['glat'].set_axislabel('Galactic Latitude')
+            except KeyError:
+                ax.coords['ra'].set_axislabel('Right Ascension')
+                ax.coords['dec'].set_axislabel('Declination')
+            except AttributeError:
+                log.info("Can't set coordinate axes. No WCS information available.")
+
+            return fig, ax, cbar
+
+        elif self.geom.ndim == 3:
+
+            try:
+                from ipywidgets.widgets.interaction import interact, fixed
+                import ipywidgets as widgets
+            except:
+                raise ImportError('ipywidgets is not installed.')
+
+            @interact(
+                     index=widgets.IntSlider(min=0, max=self.data.shape[0] - 1, step=1, value=1,
+                                             description=self.geom.axes_names[0]+' slice'),
+                     stretch=widgets.RadioButtons(options=['linear', 'sqrt', 'log'], value='sqrt',
+                                                  description='Plot stretch'),
+                     mapND=fixed(self)
+                    )
+            def _plot_interactive(mapND, index, stretch='linear'):
+                """
+                Plot ND array on matplotlib WCS axes with interactive widgets
+
+                index : integer
+                    Slice index of the non spatial dimension.
+                stretch : str
+                    Passed to `astropy.visualization.simple_norm`
+                Returns
+                -------
+                #TODO: currently returns nothing.
+                        Not sure if interact can return something.
+                        Was producing errors when I tried.
```

Comment:
I think for every call to the callback function a new `Axes` object is created, so I guess it's not possible to return those.

Reply:
agreed. When I tried it was producing weird errors